tensorflow/dtensor/mlir/collectives.cc

```diff
@@ -151,6 +151,78 @@ StatusOr<const mlir::Value> EmitAllScatter(
   return all_scatter.getOutput();
 }
 
+bool CanUseAllToAll(const dtensor::Layout& src_layout,
+                    const dtensor::Layout& tgt_layout) {
+  // All-to-all can be used for relayout if one dimension is becoming more
+  // sharded while another is becoming less sharded, for example x,unsharded ->
+  // unsharded,x.
+  // TODO(trevor-m): There may be more types of relayouts which can utilize
+  // all-to-all in addition to these which can be supported later.
+  int num_split_dims = 0;
+  int num_concat_dims = 0;
+  ShardingSpec split_spec;
+  ShardingSpec concat_spec;
+  for (int i = 0; i < src_layout.rank(); ++i) {
+    if (src_layout.sharding_spec(i) == tgt_layout.sharding_spec(i)) continue;
```

Comment:
Looks like this was not covered by the tests? Our internal tooling says if this line is changed to `if (input_layout.sharding_spec(i) != output_layout.sharding_spec(i)) continue`, (thus this functions almost always returns false) the tests still pass.

Reply:
We can probably add a test next to test_spmd_relayout_on_sparse_tensors in tensorflow/dtensor/mlir/tests/spmd_expansion.mlir

Reply:
If this function returns false, the original all-gather+scatter relayout implementation will be used so it makes sense that the tests still pass. I can add a test to make sure AllToAll is being used for relayout when we expected it to.